feathr_project/feathr/utils/_envvariableutil.py

```diff
@@ -24,51 +29,59 @@ def get_environment_variable_with_default(self, *args):
         # make it work for lower case and upper case.
         env_variable = os.environ.get(
             env_keyword, os.environ.get(upper_env_keyword))
+
+        # If the key is set in the envrionment variable, Feathr will use the value of that environment variable
         if env_variable:
             return env_variable
 
-        # if the config path doesn't exist, just return
-        try:
-            assert os.path.exists(os.path.abspath(self.config_path))
-        except:
-            logger.info("{} is not set and configuration file {} cannot be found. One of those should be set." , env_keyword, self.config_path)
+        # If it's not set in the environment, then a default is retrieved from from the feathr_config.yaml file with the same config key.
+        if os.path.exists(os.path.abspath(self.config_path)):
+            with open(os.path.abspath(self.config_path), 'r') as stream:
+                try:
+                    yaml_config = yaml.safe_load(stream)
+                    # concat all layers and  check in environment variable
+                    yaml_layer = yaml_config
 
-        with open(os.path.abspath(self.config_path), 'r') as stream:
+                    # resolve one layer after another
+                    for arg in args:
+                        yaml_layer = yaml_layer[arg]
+                    return yaml_layer
+                except KeyError as exc:
+                    logger.warning("{} not found in the config file, loading it in key vault.", env_keyword)
+                    # if not found in the config file, use key vault to get the value
+                    # return self.akv_client.get_akv_secret(env_keyword)  if self.akv_name else ""
+                except yaml.YAMLError as exc:
+                    logger.warning(exc)
+        # If it's not available in the feathr_config.yaml file, Feathr will try to reterive the value from key vault
+        if self.akv_name:
             try:
-                yaml_config = yaml.safe_load(stream)
-                # concat all layers
-                # check in environment variable
-                yaml_layer = yaml_config
-
-                # resolve one layer after another
-                for arg in args:
-                    yaml_layer = yaml_layer[arg]
-                return yaml_layer
-            except KeyError as exc:
-                logger.info(exc)
+                return  self.akv_client.get_feathr_akv_secret(env_keyword) 
+            except ResourceNotFoundError:
+                # print out warning message if cannot find the env variable in all the resources
+                logger.warning('Environment variable {} not found in environment variable, default YAML config file, or key vault service.', env_keyword)
```

Comment:
>warning [](http://example.com/codeflow?start=23&length=7)

For required enivronment vars, if the value is not found here, shoulg log error instead and stop running

Reply:
Ideally we should do so, but currently we load all the variables even if it's optional. This is worth a separate PR to solve though.